sensor/common/pubsub/lane/default.go

```diff
@@ -0,0 +1,145 @@
+
+import (
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/errorhelpers"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/sensor/common/pubsub"
+	"github.com/stackrox/rox/sensor/common/pubsub/consumer"
+	pubsubErrors "github.com/stackrox/rox/sensor/common/pubsub/errors"
+)
+
+type DefaultConfig struct {
+	Config
+}
+
+func WithDefaultLaneSize(size int) pubsub.LaneOption {
+	return func(lane pubsub.Lane) {
+		laneImpl, ok := lane.(*defaultLane)
+		if !ok {
+			return
+		}
+		laneImpl.size = size
+	}
+}
+
+func WithConsumer(consumer pubsub.NewConsumer, opts ...pubsub.ConsumerOption) pubsub.LaneOption {
+	return func(lane pubsub.Lane) {
+		laneImpl, ok := lane.(*defaultLane)
+		if !ok {
+			return
+		}
+		laneImpl.newConsumerFn = consumer
+		laneImpl.consumerOpts = opts
+	}
+}
+
+func NewDefaultLane(id pubsub.LaneID, opts ...pubsub.LaneOption) *DefaultConfig {
+	return &DefaultConfig{
+		Config: Config{
+			id:          id,
+			opts:        opts,
+			newConsumer: consumer.NewDefaultConsumer,
+		},
+	}
+}
+
+func (c *DefaultConfig) NewLane() pubsub.Lane {
+	lane := &defaultLane{
+		Lane: Lane{
+			id:            c.LaneID(),
+			newConsumerFn: c.newConsumer,
+			consumers:     make(map[pubsub.Topic][]pubsub.Consumer),
+		},
+		stopper: concurrency.NewStopper(),
+	}
+	for _, opt := range c.opts {
+		opt(lane)
+	}
+	lane.ch = make(chan pubsub.Event, lane.size)
+	go lane.run()
+	return lane
+}
+
+type defaultLane struct {
+	Lane
+	mu      sync.Mutex
+	size    int
+	ch      chan pubsub.Event
+	stopper concurrency.Stopper
+}
+
+func (l *defaultLane) Publish(event pubsub.Event) error {
+	// We need to lock here and nest two selects to avoid races stopping and
+	// publishing events
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	select {
+	case <-l.stopper.Flow().StopRequested():
+		return pubsubErrors.NewPublishOnStoppedLaneErr(l.id)
+	default:
+		return l.publish(event)
+	}
+}
+
+func (l *defaultLane) publish(event pubsub.Event) error {
+	select {
+	case <-l.stopper.Flow().StopRequested():
+		return pubsubErrors.NewPublishOnStoppedLaneErr(l.id)
+	case l.ch <- event:
+		return nil
+	}
+}
+
+func (l *defaultLane) run() {
+	defer l.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-l.stopper.Flow().StopRequested():
+			return
+		case event, ok := <-l.ch:
+			if !ok {
+				return
+			}
+			if err := l.handleEvent(event); err != nil {
+				log.Errorf("unable to handle event: %v", err)
+			}
+		}
+	}
+}
+
+func (l *defaultLane) handleEvent(event pubsub.Event) error {
+	l.consumerLock.RLock()
+	defer l.consumerLock.RUnlock()
+	consumers, ok := l.consumers[event.Topic()]
+	if !ok {
+		return pubsubErrors.NewConsumersNotFoundForTopicErr(event.Topic(), l.id)
+	}
+	errList := errorhelpers.NewErrorList("handle event")
+	for _, c := range consumers {
+		select {
+		case err := <-c.Consume(l.stopper.Client().Stopped(), event):
+			if err != nil {
+				errList.AddErrors(pubsubErrors.WrapConsumeErr(err, event.Topic(), l.id))
```

Comment:
This line is our error handling strategy for misbehaving consumers. I think that it may be to little to cover all the variety of consumers. The default behavior can be pretty noisy.

I am not sure what would be the good solution to that, but there are few variants to consider:
- Make the consumer to handle it's own error (second callback)
- Put a per-topic-worker between the lane and the consumer with some retry budget
- (radical) Define rules when misbehaving consumers are deregistered and giving them chance to register again?